File: 416-partition-equal-subset-sum/partition-equal-subset-sum.cpp
class Solution {
public:

    int backtrack(vector<int>& nums, int total, int currsum,int index, vector<vector<int>>& memo) {
        if(memo[currsum][index]!=-1) return memo[currsum][index];

        if(currsum==total/2)
        {
            memo[currsum][index]=1;
            return 1;
        }

        if (currsum>total/2 || index==nums.size()) {
            memo[currsum][index]=0;
            return 0;
        }

        int first = backtrack(nums,total,currsum,index+1,memo);
        int second = backtrack(nums,total,currsum+nums[index],index+1,memo);

        memo[currsum][index] = first || second;
        return memo[currsum][index];
    }


    bool canPartition(vector<int>& nums) {
        int n = nums.size();
        int total=0;
        for(int x : nums) total+=x;
        
        if(total%2!=0) return false;
        vector<vector<int>> memo(total+1,vector<int>(n+1,-1));
        
        int ans = backtrack(nums, total, 0, 0, memo);
        return ans == 1;
    }
};

File: 416-partition-equal-subset-sum/partition-equal-subset-sum.java
class Solution {
    public boolean canPartition(int[] nums) {
        int sum = 0;
        int n = nums.length;
        for(int x : nums) sum+= x;
        if(sum%2!=0) return false;
        int target = sum/2;
        boolean[] dp = new boolean[target+1];
        dp[0] = true;
        for(int i=1;i<=n;i++)
        {
            for(int j=target;j>=nums[i-1];j--)
            {

                dp[j] = dp[j] || dp[j-nums[i-1]];

            }
        }
        return dp[target];
    }
}